src/compute/src/logging/compute.rs

```diff
@@ -890,3 +901,74 @@ impl Drop for RetractImportFrontiers {
         self.log();
     }
 }
+
+/// Extension trait to attach `ComputeEvent::DataflowError` logging operators to collections and
+/// batch streams.
+pub(crate) trait LogDataflowErrors {
+    fn log_dataflow_errors(self, logger: Logger, export_id: GlobalId) -> Self;
+}
+
+impl<G, D> LogDataflowErrors for Collection<G, D, Diff>
+where
+    G: Scope,
+    D: Data,
+{
+    fn log_dataflow_errors(self, logger: Logger, export_id: GlobalId) -> Self {
+        self.inner
+            .unary(Pipeline, "LogDataflowErrors", |_cap, _info| {
+                let mut buffer = Vec::new();
+                move |input, output| {
+                    input.for_each(|cap, data| {
+                        data.swap(&mut buffer);
+
+                        let diff = buffer.iter().map(|(_d, _t, r)| r).sum();
+                        logger.log(ComputeEvent::ErrorCount { export_id, diff });
+
+                        output.session(&cap).give_vec(&mut buffer);
+                    });
+                }
+            })
+            .as_collection()
+    }
+}
+
+impl<G, B> LogDataflowErrors for Stream<G, B>
+where
+    G: Scope,
+    B: BatchReader<R = Diff> + Clone + 'static,
+{
+    fn log_dataflow_errors(self, logger: Logger, export_id: GlobalId) -> Self {
+        self.unary(Pipeline, "LogDataflowErrors", |_cap, _info| {
+            let mut buffer = Vec::new();
+            move |input, output| {
+                input.for_each(|cap, data| {
+                    data.swap(&mut buffer);
+
+                    let diff = buffer.iter().map(sum_batch_diffs).sum();
+                    logger.log(ComputeEvent::ErrorCount { export_id, diff });
+
+                    output.session(&cap).give_vec(&mut buffer);
+                });
+            }
+        })
+    }
+}
+
+/// Return the sum of all diffs within the given batch.
+fn sum_batch_diffs<B>(batch: &B) -> Diff
```

Comment:
We might want to mention this is is `O(unique (value, key, time))`, so it can be slow for large collections. I don't think it's a problem since we only attach it to error streams, but we might want to add a note to our future selves that we don't attach it to non-error streams.